core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,399 @@
+ * advanced TLS features, such as root certificate reloading, peer cert custom verification, etc.
+ */
+@IgnoreJRERequirement
+public final class AdvancedTlsX509TrustManager extends X509ExtendedTrustManager {
+  private static final Logger log = Logger.getLogger(AdvancedTlsX509TrustManager.class.getName());
+
+  private Verification verification;
+  private PeerVerifier peerVerifier;
+  private SslSocketPeerVerifier sslSocketPeerVerifier;
+  private SslEnginePeerVerifier sslEnginePeerVerifier;
+  // The trust certs we will use to verify peer certificate chain.
+  private volatile X509Certificate[] trustCerts;
+
+  private AdvancedTlsX509TrustManager(Verification verification,  PeerVerifier peerVerifier,
+      SslSocketPeerVerifier socketPeerVerifier, SslEnginePeerVerifier enginePeerVerifier) {
+    this.verification = verification;
+    this.peerVerifier = peerVerifier;
+    this.sslSocketPeerVerifier = socketPeerVerifier;
+    this.sslEnginePeerVerifier = enginePeerVerifier;
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, false);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain,  String authType, SSLEngine engine)
+      throws CertificateException {
+    checkTrusted(chain, authType, engine, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, null, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+      throws CertificateException {
+    checkTrusted(chain, authType, null, socket, true);
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return trustCerts;
+  }
+
+  // If this is set, we will use the default trust certificates stored on user's local system.
+  // After this is used, functions that will update this.trustCerts(e.g. updateTrustCredentials(),
+  // updateTrustCredentialsFromFile()) should not be called.
+  public void useSystemDefaultTrustCerts() {
+    this.trustCerts = new X509Certificate[0];
+  }
+
+  /**
+   * Updates the current cached trust certificates.
+   *
+   * @param trustCerts  the trust certificates that are going to be used
+   */
+  public void updateTrustCredentials(X509Certificate[] trustCerts) {
+    // Question: in design, we will do an copy of trustCerts, but that seems not easy, since
```

Comment:
Copy the certificate or the array?

Reply:
Oh yeah, I meant copying the certificate seemed hard. Can we use the same reference of the certificate, but each time create a new copy of the array? If we can, then I think we are good.

Reply:
Awesome! Resolving this...